WorkspaceServer/Servers/Roslyn/TrackingStringWriter.cs

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace WorkspaceServer.Servers.Roslyn
+{
+    internal class TrackingStringWriter : StringWriter
+    {
+        private class Region
+        {
+            public int Start { get; set; }
+            public int Length { get; set; }
+        }
+         List<Region> _regions = new List<Region>();
+         private bool _trakingSize;
+
+
+         public bool WriteOccurred { get; set; }
+
+        public override void Write(char value)
+        {
+            TrackSize(() => base.Write(value));
+        }
+
+        private void TrackSize(Action action)
+        {
+            WriteOccurred = true;
+            if (_trakingSize)
+            {
+                action();
+                return;
+            }
+
+            _trakingSize = true;
+            var sb = base.GetStringBuilder();
+
+            var region = new Region
+            {
+                Start = sb.Length
+            };
+
+            _regions.Add(region);
+            action();
+            region.Length = sb.Length - region.Start;
+            _trakingSize = false;
+        }
+        private async Task TrackSizeAsync(Func<Task> action)
+        {
+            WriteOccurred = true;
+            if (_trakingSize)
```

Comment:
I feel like this is trying to handle some concurrency scenario but there doesn't seem to be any synchronization?

Reply:
is handling re-entrancy, writeLine get decombosed in base class in write(value) and then WriteLine()